sdk/core/Azure.Core.Experimental/samples/Variant.md

```diff
@@ -0,0 +1,135 @@
+
+- [Assigning a value to `Variant`](#assign-a-value-to-variant)
+- [Retrieving the value a `Variant` holds](#get-the-value-from-variant)
+- [Working with `null` and `Variant`](#handling-nulls)
+
+## Assign a value to Variant
+
+`Variant` has implicit cast operators for each of the primitives listed above.  This means you can assign a value of those types to `Variant` without casting, and a new `Variant` instance holding an unboxed copy of the value will be created.
+
+```csharp
+Variant v = 3L;
+
+// v.Type is System.Int64
+```
+
+Since you can't have an implicit cast operator from `object`, for reference types, you must create a new instance of `Variant` and pass the value to the `Variant` constructor.
+
+```csharp
+MemoryStream s = new();
+Variant v = new(s);
+
+// v.Type is System.IO.MemoryStream
+```
+
+`Variant` will box value types that aren't on the list above, because it only stores 16 bytes and user-defined value types might exceed this size.  The one exception is for `enum`, which `Variant` can hold without boxing.  Because we can't add implicit cast operators for unknown enum types, you must call the `Create` method to create a `Variant` that holds an enum without boxing.
+
+```csharp
+Value v = Value.Create(System.Color.Blue);
+
+// v.Type is System.Color
+```
+
+## Get the value from Variant
+
+`Variant` has explicit cast operators for each of the primitives listed above, as well as to `string`.  That means you can assign a `Variant` holding one of these types to a variable of that type with a cast:
+
+```csharp
+Variant v = 3;
+int i = (int)v;
+```
+
+If you don't know the type of the value that a `Variant` is holding, you can read that from the `Variant.Type` property.
+
+```csharp
+switch (v.Type)
+{
+    case Type s when s == typeof(string):
+        Console.WriteLine($"string: {v}");
+        break;
+    case Type i when i == typeof(int):
+        Console.WriteLine($"int: {(int)v}");
+        break;
+}
+```
+
+If you do know the type the `Variant` is holding, you can use the `As<T>` method to get its value as that type.
+
+```csharp
+MemoryStream s = new();
+Variant v = new (s);
+
+MemoryStream streamValue = v.As<MemoryStream>();
+```
+
+If you call `Variant.As<T>` and the `Variant` isn't holding the type that you ask for, it will throw an `InvalidCastException`.  To try to retrieve the value without risking an exception being thrown, you can use the `TryGetValue` method instead.
+
+```csharp
+if (v.TryGetValue(out string s))
+{
+    Console.WriteLine($"string: {s}");
+}
+
+if (v.TryGetValue(out int i))
+{
+    Console.WriteLine($"int: {i}");
+}
+```
+
+## Handling nulls
+
+`Variant` handles nullable primitives without boxing them for value types on the supported list above.  It also supports holding a reference type with a `null` value.  In either of these cases, the variant's `Type` property will return `null`.
```

Comment:
The way I read this, any nullable value type will have a `Type` property value of `null`. Is that correct, or is it only when the value of the nullable value type is currently null?

Reply:
Ah, good.  Let me clarify this.  IIRC, the Type property is only null for a nullable value type when the value of the Nullable is null.

Before I checked, I had [originally written](https://github.com/annelo-msft/gists/blob/main/variant/api.md#handling-nulls) that for value types we would return the actual type from the Type property (e.g. `System.Nullable<System.Int32>`, but for reference types we don't.  I think this would be easy to implement but we don't do it today.  We aren't able to store the type for reference types because it's either entirely unknown or would exceed our 128 bit storage limit.  

What do you think?  Would that be a better API experience?

Reply:
I do think it would be much better to return the actual nullable type, if possible. The alternatives both seem worse (e.g. `Type`'s value can change depending on the value, or all nullable value types return `null`)

Reply:
Thinking to COM's `VARIANT`, though, this is analogous to `VT_NULL`. Sure, it's an `int?`, but because it's null does it even have a type? And what can any caller do with it besides treat it as null? Or do you feel there's sufficient coverage with the `IsNull` property?

Reply:
> I do think it would be much better to return the actual nullable type, if possible.

I agree with this.  Let me look at a) how hard it is to implement and b) what the implications are for type switching.  I think in cases where we come in through a specific constructor (e.g. `int? i = null; V v= i;`), we can be clear about the type, and even though the value is null, if someone tried to do `v = new MemoryStream()` we could yell at them for type switching.  It feels like there is a user intent here that we could support through the API.  I'm not deep on `VARIANT`, but I know that people say it causes them nausea, so I don't think we have to preserve those rules here if we think there are better APIs to follow in a typesafe world.

I'm happy to look at examples where this gets interesting though, so we can think about whether a design choice like this might paint us into a corner.  What are the pro/con cases here?